File: realm/realm-library/src/androidTest/java/io/realm/SyncConfigurationTests.java
/*
 * Copyright 2016 Realm Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.realm;

import android.content.Context;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.junit.runner.RunWith;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

import io.realm.rule.RunInLooperThread;
import io.realm.rule.TestRealmConfigurationFactory;

import static io.realm.util.SyncTestUtils.createTestUser;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

@RunWith(AndroidJUnit4.class)
public class SyncConfigurationTests {
    @Rule
    public final TestRealmConfigurationFactory configFactory = new TestRealmConfigurationFactory();

    @Rule
    public final RunInLooperThread looperThread = new RunInLooperThread();

    @Rule
    public final TemporaryFolder tempFolder = new TemporaryFolder();

    private Context context;

    @Before
    public void setUp() {
        context = InstrumentationRegistry.getContext();
    }

    @After
    public void tearDown() throws Exception {
    }

    @Test
    public void user() {
//        new SyncConfiguration.Builder(context);
        // Check that user can be added
        // That the default local path is correct
    }

    @Test
    public void user_invalidUserThrows() {
        SyncConfiguration.Builder builder = new SyncConfiguration.Builder(context);

        try {
            builder.user(null);
        } catch (IllegalArgumentException ignore) {
        }

        User user = createTestUser(0); // Create user that has expired credentials
        try {
            builder.user(user);
        } catch (IllegalArgumentException ignore) {
        }
    }

    @Test
    public void serverUrl_setsFolderAndFileName() {
        User user = createTestUser();
        String[][] validUrls = {
                // <URL>, <Folder>, <FileName>
                { "realm://objectserver.realm.io/~/default", "realm-object-server/" + user.getIdentity(), "default" },
                { "realm://objectserver.realm.io/~/sub/default", "realm-object-server/" + user.getIdentity() + "/sub", "default" }
        };

        for (String[] validUrl : validUrls) {
            String serverUrl  = validUrl[0];
            String expectedFolder = validUrl[1];
            String expectedFileName = validUrl[2];

            SyncConfiguration config = new SyncConfiguration.Builder(context)
                    .serverUrl(serverUrl)
                    .user(user)
                    .build();

            assertEquals(new File(context.getFilesDir(), expectedFolder), config.getRealmDirectory());
            assertEquals(expectedFileName, config.getRealmFileName());
        }
    }

    @Test
    public void serverUrl_invalidUrlThrows() {
        String[] invalidUrls = {
            null,
// TODO Should these two fail?
//            "objectserver.realm.io/~/default", // Missing protocol. TODO Should we just default to one?
//            "/~/default", // Missing server
            "realm://objectserver.realm.io/~/default.realm", // Ending with .realm
            "realm://objectserver.realm.io/~/default.realm.lock", // Ending with .realm.lock
            "realm://objectserver.realm.io/~/default.realm.management", // Ending with .realm.management
            "realm://objectserver.realm.io/<~>/default.realm", // Invalid chars <>
            "realm://objectserver.realm.io/~/default.realm/", // Ending with /
            "realm://objectserver.realm.io/~/Αθήνα", // Non-ascii
            "realm://objectserver.realm.io/~/foo/../bar", // .. is not allowed
            "realm://objectserver.realm.io/~/foo/./bar", // . is not allowed
            "http://objectserver.realm.io/~/default", // wrong scheme
        };

        SyncConfiguration.Builder builder = new SyncConfiguration.Builder(context);
        for (String invalidUrl : invalidUrls) {
            try {
                builder.serverUrl(invalidUrl);
                fail(invalidUrl + " should have failed.");
            } catch (IllegalArgumentException ignore) {
            }
        }
    }

    private String makeServerUrl(int len) {
        StringBuilder builder = new StringBuilder("realm://objectserver.realm.io/~/");
        for (int i = 0; i < len; i++) {
            builder.append('A');
        }
        return builder.toString();
    }

    @Test
    public void serverUrl_length() {
        int[] lengths = {1, SyncConfiguration.MAX_FILE_NAME_LENGTH - 1,
                SyncConfiguration.MAX_FILE_NAME_LENGTH, SyncConfiguration.MAX_FILE_NAME_LENGTH + 1, 1000};

        for (int len : lengths) {
            SyncConfiguration.Builder builder = new SyncConfiguration.Builder(context)
                .serverUrl(makeServerUrl(len))
                .user(createTestUser());

            SyncConfiguration config = builder.build();
            assertTrue("Length: " + len, config.getRealmFileName().length() <= SyncConfiguration.MAX_FILE_NAME_LENGTH);
            assertTrue("Length: " + len, config.getPath().length() <= SyncConfiguration.MAX_FULL_PATH_LENGTH);
        }
    }

    @Test
    public void serverUrl_invalidChars() {
        SyncConfiguration.Builder builder = new SyncConfiguration.Builder(context)
                .serverUrl("realm://objectserver.realm.io/~/?")
                .user(createTestUser());
        SyncConfiguration config = builder.build();
        assertFalse(config.getRealmFileName().contains("?"));
    }

    @Test
    public void serverUrl_port() {
        Map<String, Integer> urlPort = new HashMap<String, Integer>();
        urlPort.put("realm://objectserver.realm.io/~/default", SyncConfiguration.PORT_REALM);
        urlPort.put("realms://objectserver.realm.io/~/default", SyncConfiguration.PORT_REALMS);
        urlPort.put("realm://objectserver.realm.io:8080/~/default", 8080);
        urlPort.put("realms://objectserver.realm.io:2443/~/default", 2443);

        for (String url : urlPort.keySet()) {
            SyncConfiguration config = new SyncConfiguration.Builder(context)
                    .serverUrl(url)
                    .user(createTestUser())
                    .build();
            assertEquals(urlPort.get(url).intValue(), config.getServerUrl().getPort());
        }
    }

    @Test
    public void userAndServerUrlRequired() {
        SyncConfiguration.Builder builder;

        // Both missing
        builder = new SyncConfiguration.Builder(context);
        try {
            builder.build();
        } catch (IllegalStateException ignore) {
        }

        builder = new SyncConfiguration.Builder(context);
        try {
            builder.user(createTestUser(Long.MAX_VALUE)).build();
        } catch (IllegalStateException ignore) {
        }

        // user missing
        builder = new SyncConfiguration.Builder(context);
        try {
            builder.serverUrl("realm://foo.bar/~/default").build();
        } catch (IllegalStateException ignore) {
        }
    }

    @Test
    public void errorHandler() {
        SyncConfiguration.Builder builder;
        builder = new SyncConfiguration.Builder(context)
                .user(createTestUser())
                .serverUrl("realm://objectserver.realm.io/default");

        Session.ErrorHandler errorHandler = new Session.ErrorHandler() {
            @Override
            public void onError(Session session, ObjectServerError error) {

            }
        };

        SyncConfiguration config = builder.errorHandler(errorHandler).build();
        assertEquals(errorHandler, config.getErrorHandler());
    }

    @Test
    public void errorHandler_fromSyncManager() {
        // Set default error handler
        Session.ErrorHandler errorHandler = new Session.ErrorHandler() {
            @Override
            public void onError(Session session, ObjectServerError error) {

            }
        };
        SyncManager.setDefaultSessionErrorHandler(errorHandler);

        // Create configuration using the default handler
        SyncConfiguration config = new SyncConfiguration.Builder(context)
                .user(createTestUser())
                .serverUrl("realm://objectserver.realm.io/default")
                .build();
        assertEquals(errorHandler, config.getErrorHandler());
        SyncManager.setDefaultSessionErrorHandler(null);
    }


    @Test
    public void errorHandler_nullThrows() {
        SyncConfiguration.Builder builder;
        builder = new SyncConfiguration.Builder(context);

        try {
            builder.errorHandler(null);
        } catch (IllegalArgumentException ignore) {
        }
    }

}

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SyncConfigurationTests.kt
/*
 * Copyright 2020 Realm Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.realm.mongodb.sync

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import io.realm.*
import io.realm.coroutines.FlowFactory
import io.realm.entities.SyncStringOnly
import io.realm.entities.SyncStringOnlyModule
import io.realm.kotlin.createObject
import io.realm.kotlin.where
import io.realm.mongodb.AppException
import io.realm.mongodb.SyncTestUtils.Companion.createTestUser
import io.realm.mongodb.User
import io.realm.mongodb.close
import io.realm.mongodb.registerUserAndLogin
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOf
import org.bson.BsonString
import org.bson.types.ObjectId
import org.junit.*
import org.junit.runner.RunWith
import kotlin.test.*

@RunWith(AndroidJUnit4::class)
class SyncConfigurationTests {

    companion object {
        private const val DEFAULT_PARTITION = "default"
    }

    @get:Rule
    val configFactory = TestSyncConfigurationFactory()

    private lateinit var app: TestApp

    @Before
    fun setUp() {
        Realm.init(InstrumentationRegistry.getInstrumentation().targetContext)
        app = TestApp()
    }

    @After
    fun tearDown() {
        if (this::app.isInitialized) {
            app.close()
        }
    }

    @Test
    fun errorHandler() {
        val builder: SyncConfiguration.Builder = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
        val errorHandler: SyncSession.ErrorHandler = object : SyncSession.ErrorHandler {
            override fun onError(session: SyncSession, error: AppException) {}
        }
        val config = builder.errorHandler(errorHandler).build()
        assertEquals(errorHandler, config.errorHandler)
    }

    @Test
    fun errorHandler_fromAppConfiguration() {
        val user: User = createTestUser(app)
        val config: SyncConfiguration = SyncConfiguration.defaultConfig(user, DEFAULT_PARTITION)
        assertEquals(app.configuration.defaultErrorHandler, config.errorHandler)
    }

    @Test
    fun errorHandler_nullThrows() {
        val user: User = createTestUser(app)
        val builder = SyncConfiguration.Builder(user, DEFAULT_PARTITION)
        assertFailsWith<IllegalArgumentException> { builder.errorHandler(TestHelper.getNull()) }
    }

    @Test
    fun clientResetHandler() {
        val builder: SyncConfiguration.Builder = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
        val handler = object : SyncSession.ClientResetHandler {
            override fun onClientReset(session: SyncSession, error: ClientResetRequiredError) {}
        }
        val config = builder.clientResetHandler(handler).build()
        assertEquals(handler, config.clientResetHandler)
    }

    @Test
    fun clientResetHandler_fromAppConfiguration() {
        val user: User = createTestUser(app)
        val config: SyncConfiguration = SyncConfiguration.defaultConfig(user, DEFAULT_PARTITION)
        assertEquals(app.configuration.defaultClientResetHandler, config.clientResetHandler)
    }

    @Test
    fun clientResetHandler_nullThrows() {
        val user: User = createTestUser(app)
        val builder = SyncConfiguration.Builder(user, DEFAULT_PARTITION)
        assertFailsWith<IllegalArgumentException> { builder.clientResetHandler(TestHelper.getNull()) }
    }

    @Test
    fun equals() {
        val user: User = createTestUser(app)
        val config: SyncConfiguration = SyncConfiguration.defaultConfig(user, DEFAULT_PARTITION)
        assertTrue(config == config)
    }

    @Test
    fun equals_same() {
        val user: User = createTestUser(app)
        val config1: SyncConfiguration = SyncConfiguration.Builder(user, DEFAULT_PARTITION).build()
        val config2: SyncConfiguration = SyncConfiguration.Builder(user, DEFAULT_PARTITION).build()
        assertTrue(config1 == config2)
    }

    @Test
    // FIXME Tests are not exhaustive
    fun equals_not() {
        val user1: User = createTestUser(app)
        val user2: User = createTestUser(app)
        val config1: SyncConfiguration = SyncConfiguration.Builder(user1, DEFAULT_PARTITION).build()
        val config2: SyncConfiguration = SyncConfiguration.Builder(user2, DEFAULT_PARTITION).build()
        assertFalse(config1 == config2)
    }

    @Test
    fun hashCode_equal() {
        val user: User = createTestUser(app)
        val config: SyncConfiguration = SyncConfiguration.defaultConfig(user, DEFAULT_PARTITION)
        assertEquals(config.hashCode(), config.hashCode())
    }

    @Test
    fun hashCode_notEquals() {
        val user1: User = createTestUser(app)
        val user2: User = createTestUser(app)
        val config1: SyncConfiguration = SyncConfiguration.defaultConfig(user1, DEFAULT_PARTITION)
        val config2: SyncConfiguration = SyncConfiguration.defaultConfig(user2, DEFAULT_PARTITION)
        assertNotEquals(config1.hashCode(), config2.hashCode())
    }

    @Test
    fun get_syncSpecificValues() {
        val user: User = createTestUser(app)
        val config: SyncConfiguration = SyncConfiguration.defaultConfig(user, DEFAULT_PARTITION)
        assertTrue(user == config.user)
        assertEquals("ws://127.0.0.1:9090/", config.serverUrl.toString()) // FIXME: Figure out exactly what to return here
        assertFalse(config.shouldDeleteRealmOnLogout())
        assertTrue(config.isSyncConfiguration)
    }

    @Test
    fun name() {
        val user: User = createTestUser(app)
        val filename = "my-file-name.realm"
        val config: SyncConfiguration = SyncConfiguration.Builder(user, DEFAULT_PARTITION)
                .name(filename)
                .build()
        val suffix = "/mongodb-realm/${user.app.configuration.appId}/${user.id}/$filename"
        assertTrue(config.path.endsWith(suffix))
    }

    @Test
    fun name_illegalValuesThrows() {
        val user: User = createTestUser(app)
        val builder = SyncConfiguration.Builder(user, DEFAULT_PARTITION)

        assertFailsWith<IllegalArgumentException> { builder.name(TestHelper.getNull()) }
        assertFailsWith<IllegalArgumentException> { builder.name(".realm") }
    }

    @Test
    fun encryption() {
        val user: User = createTestUser(app)
        val config: SyncConfiguration = SyncConfiguration.Builder(user, DEFAULT_PARTITION)
                .encryptionKey(TestHelper.getRandomKey())
                .build()
        assertNotNull(config.encryptionKey)
    }

    @Test
    fun encryption_invalid_null() {
        val user: User = createTestUser(app)
        val builder = SyncConfiguration.Builder(user, DEFAULT_PARTITION)
        assertFailsWith<IllegalArgumentException> { builder.encryptionKey(TestHelper.getNull()) }
    }

    @Test
    fun encryption_invalid_wrong_length() {
        val user: User = createTestUser(app)
        val builder = SyncConfiguration.Builder(user, DEFAULT_PARTITION)
        assertFailsWith<IllegalArgumentException> { builder.encryptionKey(byteArrayOf(1, 2, 3)) }
    }

    @Test
    fun initialData() {
        val user: User = createTestUser(app)
        val config = configFactory.createSyncConfigurationBuilder(user)
                .schema(SyncStringOnly::class.java)
                .initialData(object : Realm.Transaction {
                    override fun execute(realm: Realm) {
                        val stringOnly: SyncStringOnly = realm.createObject(ObjectId())
                        stringOnly.chars = "TEST 42"
                    }
                })
                .build()
        assertNotNull(config.initialDataTransaction)

        // open the first time - initialData must be triggered
        Realm.getInstance(config).use { realm ->
            val results: RealmResults<SyncStringOnly> = realm.where<SyncStringOnly>().findAll()
            assertEquals(1, results.size)
            assertEquals("TEST 42", results.first()!!.chars)
        }

        // open the second time - initialData must not be triggered
        Realm.getInstance(config).use { realm ->
            assertEquals(1, realm.where<SyncStringOnly>().count())
        }
    }

    @Test
    fun defaultRxFactory() {
        val user: User = createTestUser(app)
        val config: SyncConfiguration = SyncConfiguration.defaultConfig(user, DEFAULT_PARTITION)
        assertNotNull(config.rxFactory)
    }

    @Test
    fun toString_nonEmpty() {
        val user: User = createTestUser(app)
        val config: SyncConfiguration = SyncConfiguration.defaultConfig(user, DEFAULT_PARTITION)
        val configStr = config.toString()
        assertTrue(configStr.isNotEmpty())
    }

    // Check that it is possible for multiple users to reference the same Realm URL while each user still use their
    // own copy on the filesystem. This is e.g. what happens if a Realm is shared using a PermissionOffer.
    @Test
    fun multipleUsersReferenceSameRealm() {
        val user1: User = app.registerUserAndLogin(TestHelper.getRandomEmail(), "123456")
        val user2: User = app.registerUserAndLogin(TestHelper.getRandomEmail(), "123456")

        val config1: SyncConfiguration = SyncConfiguration.Builder(user1, DEFAULT_PARTITION)
                .modules(SyncStringOnlyModule())
                .build()
        val config2: SyncConfiguration = SyncConfiguration.Builder(user2, DEFAULT_PARTITION)
                .modules(SyncStringOnlyModule())
                .build()

        // Verify that two different configurations can be used for the same URL
        val realm1: Realm = Realm.getInstance(config1)
        val realm2: Realm = Realm.getInstance(config2)
        assertNotEquals(realm1, realm2)

        realm1.close()
        realm2.close()

        // Verify that we actually save two different files
        assertNotEquals(config1.path, config2.path)
    }

    @Test
    fun defaultConfiguration_throwsIfNotLoggedIn() {
        // TODO Maybe we could avoid registering a real user
        val user = app.registerUserAndLogin(TestHelper.getRandomEmail(), "123456")
        user.logOut()
        assertFailsWith<IllegalArgumentException> { SyncConfiguration.defaultConfig(user, DEFAULT_PARTITION) }
    }

    @Test
    @Ignore("Not implemented yet")
    fun shouldWaitForInitialRemoteData() {
    }

    @Test
    @Ignore("Not implemented yet")
    fun getInitialRemoteDataTimeout() {
    }

    @Test
    @Ignore("Not implemented yet")
    fun getSessionStopPolicy() {
    }

    @Test
    @Ignore("Not implemented yet")
    fun getUrlPrefix() {
    }

    @Test
    fun getPartitionValue() {
        val user: User = createTestUser(app)
        val config: SyncConfiguration = SyncConfiguration.defaultConfig(user, DEFAULT_PARTITION)
        assertEquals(BsonString(DEFAULT_PARTITION), config.partitionValue)
    }

    @Test
    fun clientResyncMode() {
        val user: User = createTestUser(app)

        // Default mode for full Realms
        var config: SyncConfiguration = SyncConfiguration.defaultConfig(user, DEFAULT_PARTITION)
        assertEquals(ClientResyncMode.MANUAL, config.clientResyncMode)

        // Manually set the mode
        config = SyncConfiguration.Builder(user, DEFAULT_PARTITION)
                .clientResyncMode(ClientResyncMode.RECOVER_LOCAL_REALM)
                .build()
        assertEquals(ClientResyncMode.RECOVER_LOCAL_REALM, config.clientResyncMode)
    }

    @Test
    fun clientResyncMode_throwsOnNull() {
        val user: User = createTestUser(app)
        val config: SyncConfiguration.Builder = SyncConfiguration.Builder(user, DEFAULT_PARTITION)
        try {
            config.clientResyncMode(TestHelper.getNull())
            fail()
        } catch (ignore: IllegalArgumentException) {
        }
    }

    // If the same user create two configurations with different partition values they must
    // resolve to different paths on disk.
    @Test
    fun differentPartitionValuesAreDifferentRealms() {
        val user: User = createTestUser(app)
        val config1 = SyncConfiguration.Builder(user, "realm1").modules(SyncStringOnlyModule()).build()
        val config2 = SyncConfiguration.Builder(user, "realm2").modules(SyncStringOnlyModule()).build()
        assertNotEquals(config1.path, config2.path)

        assertTrue(config1.path.endsWith("${app.configuration.appId}/${user.id}/s_realm1.realm"))
        assertTrue(config2.path.endsWith("${app.configuration.appId}/${user.id}/s_realm2.realm"))

        // Check for https://github.com/realm/realm-java/issues/6882
        val realm1 = Realm.getInstance(config1)
        try {
            val realm2 = Realm.getInstance(config2)
            realm2.close()
        } finally {
            realm1.close()
        }
    }

    @Test
    fun nullPartitionValue() {
        val user: User = createTestUser(app)

        val configs = listOf<SyncConfiguration>(
                SyncConfiguration.defaultConfig(user, null as String?),
                SyncConfiguration.defaultConfig(user, null as Int?),
                SyncConfiguration.defaultConfig(user, null as Long?),
                SyncConfiguration.defaultConfig(user, null as ObjectId?),
                SyncConfiguration.Builder(user, null as String?).build(),
                SyncConfiguration.Builder(user, null as Int?).build(),
                SyncConfiguration.Builder(user, null as Long?).build(),
                SyncConfiguration.Builder(user, null as ObjectId?).build()
        )

        configs.forEach { config ->
            assertTrue(config.path.endsWith("/null.realm"))
        }
    }

    @Test
    fun loggedOutUsersThrows() {
        val user: User = app.registerUserAndLogin(TestHelper.getRandomEmail(), "123456")
        user.logOut()
        assertFailsWith<java.lang.IllegalArgumentException> {
            SyncConfiguration.defaultConfig(user, ObjectId())
        }
        assertFailsWith<java.lang.IllegalArgumentException> {
            SyncConfiguration.defaultConfig(app.currentUser(), ObjectId())
        }
    }

    @Test
    fun allowQueriesOnUiThread_defaultsToTrue() {
        val builder: SyncConfiguration.Builder = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
        val configuration = builder.build()
        assertTrue(configuration.isAllowQueriesOnUiThread)
    }

    @Test
    fun allowQueriesOnUiThread_explicitFalse() {
        val builder: SyncConfiguration.Builder = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
        val configuration = builder.allowQueriesOnUiThread(false)
                .build()
        assertFalse(configuration.isAllowQueriesOnUiThread)
    }

    @Test
    fun allowQueriesOnUiThread_explicitTrue() {
        val builder: SyncConfiguration.Builder = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
        val configuration = builder.allowQueriesOnUiThread(true)
                .build()
        assertTrue(configuration.isAllowQueriesOnUiThread)
    }

    @Test
    fun allowWritesOnUiThread_defaultsToFalse() {
        val builder: SyncConfiguration.Builder = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
        val configuration = builder.build()
        assertFalse(configuration.isAllowWritesOnUiThread)
    }

    @Test
    fun allowWritesOnUiThread_explicitFalse() {
        val builder: SyncConfiguration.Builder = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
        val configuration = builder.allowWritesOnUiThread(false)
                .build()
        assertFalse(configuration.isAllowWritesOnUiThread)
    }

    @Test
    fun allowWritesOnUiThread_explicitTrue() {
        val builder: SyncConfiguration.Builder = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
        val configuration = builder.allowWritesOnUiThread(true)
                .build()
        assertTrue(configuration.isAllowWritesOnUiThread)
    }

    @Test
    fun coroutinesFactory_defaultNonNull() {
        val configuration = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
                .build()
        assertNotNull(configuration.flowFactory)
    }

    @Test
    fun coroutinesFactory() {
        val factory = object : FlowFactory {
            override fun from(realm: Realm): Flow<Realm> {
                return flowOf()
            }

            override fun from(realm: DynamicRealm): Flow<DynamicRealm> {
                return flowOf()
            }

            override fun <T : RealmModel?> from(realm: Realm, results: RealmResults<T>): Flow<RealmResults<T>> {
                return flowOf()
            }

            override fun <T : RealmObject?> from(realm: Realm, realmList: RealmList<T>): Flow<RealmList<T>> {
                return flowOf()
            }

            override fun <T : DynamicRealmObject?> from(realm: DynamicRealm, realmList: RealmList<T>): Flow<RealmList<T>> {
                return flowOf()
            }

            override fun <T : RealmModel?> from(realm: Realm, realmObject: T): Flow<T> {
                return flowOf()
            }

            override fun from(realm: DynamicRealm, dynamicRealmObject: DynamicRealmObject): Flow<DynamicRealmObject> {
                return flowOf()
            }
        }

        val configuration1 = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
                .flowFactory(factory)
                .build()
        assertEquals(factory, configuration1.flowFactory)

        val configuration2 = SyncConfiguration.Builder(createTestUser(app), DEFAULT_PARTITION)
                .build()
        assertNotEquals(factory, configuration2.flowFactory)
    }
}
